### ISSUE TITLE
Examples (static & interactive)

### ISSUE THREAD

**Issue by martinklepsch** [cljdoc/cljdoc#70]
We would like to allow library users and authors to attach examples to a release of their library, a namespace or a specific var. These examples could be just static code snippets with some explanation or even be interactive ([dynadoc](https://github.com/oakes/Dynadoc), [klipse](https://github.com/viebel/klipse) et al).

There are a bunch of things to think about here:

- **Ease of contribution:** as low as possible while preventing spam
- **Versioning:** examples may only apply to a var/namespace in a specific version range (or we could just be assuming people follow Accretion principles)
- **Licensing:** all examples should be contributed under one license and be easily available for download (Git?)
- **Specification:** static examples could be plain Markdown but interactive examples probably require a more structured approach. EDN? Markdown front-matter? We should come up with some specific proposals around that.

Note that this also likely depends on #58 and moving to a more flexible storage backend than Grimoire (pre-defined data model).

**EDIT:** Note that I'm mostly thinking of ClojureScript based examples here. For Clojure examples we'd need a runtime on some server and I wouldn't really want to get into that unless someone has a good proposal.

**EDIT2:** Maybe starting with just static examples would be a reasonable way to go as that covers 3 of the 4 points above.

**Comment by alex-dixon** [comment]
Re: Clojure vs. Clojurescript....How is the project's infrastructure paid for? Would Clojure examples necessarily entail significant cost given the current setup?

Should dependencies be on the list of things to think about?

**Comment by martinklepsch** [comment]
> How is the project's infrastructure paid for?

Currently it's running on AWS (5% of costs) DigitalOcean (95%), AWS is paid by me, DigitalOcean invoices are still covered by a generous credit I received from a friend working there. 

> Would Clojure examples necessarily entail significant cost given the current setup?

I don't really know but I would expect that they would require some server to run so how big that cost is probably largely depends on how efficient that backend can be used. An alternative could also be using other services like [repl.it](https://repl.it/repls/DoubleBoilingLogin) or [NextJournal](https://nextjournal.com/) but that will potentially imply some lock-in/reliance on those services etc.

> Should dependencies be on the list of things to think about?

I would expect them to, yes. Most library code won't be usable without dependencies being available I guess. I remember seeing that it's possible to dynamically load stuff into Klipse for instance but I'm not sure how easy that would be too achieve.

**Comment by alex-dixon** [comment]
@martinklepsch Cool. Focusing on static examples:

> We would like to allow library users and authors to attach examples to a release of their library, a namespace or a specific var.

Static examples are often written into docstrings, which we're already slurping in. I'm wondering what advantages we can offer library authors over that. I'd gladly go out of my way to get interactive examples on cljdoc if it required me to write something like Dynadoc's `defexample`s. I'd be less likely to do that for static examples though.

Is your idea that anyone can contribute to an "examples" section of a library's clljdoc site? Just like clojuredoc but for community libraries **and** core libraries? I think that'd be really neat.

If library authors are the only ones who can contribute examples, 1. we'd probably have fewer examples, 2. core libraries would likely need to have examples written for them or otherwise be handled as a special case.

**Comment by martinklepsch** [comment]
> Is your idea that anyone can contribute to an "examples" section of a library's clljdoc site?

Yes very much so. Also note that examples should be attachable to most entities, e.g. version, namespace, var. [1]

> Static examples are often written into docstrings, which we're already slurping in. I'm wondering what advantages we can offer library authors over that.)

We are already pulling in docstrings but they aren't very structured. If we want to open up contribution of examples as part of the docstring we would need to facilitate some kind of consensus about what examples are included. If we keep user contributed examples separately we can avoid that problem.

Opening this up for contributions is the value we offer to library authors. 

> I'd gladly go out of my way to get interactive examples on cljdoc if it required me to write something like Dynadoc's `defexample`s.

As I said, I think we can totally do this for ClojureScript but it's more tricky for Clojure. One thing I'd like to explore is a more data-driven way to specify examples. Reading them from Clojure files (as with `dynadoc`) is costly  and I wonder if edn or some kind of REPL transcript (https://github.com/cognitect-labs/transcriptor) might be better suited.

In any case I'm not intending to give up on interactive examples but rather want to get going on static ones so we have the foundations in place to properly experiment with whatever approach is taken. 

Let me know if any of this doesn't make sense 🙂 

---
[1] Maybe thinking of these in terms of attachments would be interesting and I think it's somewhere along the lines of @arrdem's work with Stacks/Grimoire. An attachment could be an example but it could also be an alternative docstring, a test case, etc.

**Comment by ajoberstar** [comment]
Dynadoc is focused on running locally; maybe cljdoc could have a similar mode. Dynamic Clojure examples could work locally, but generate static ones for cljdoc.xyz.

**Comment by martinklepsch** [comment]
@ajoberstar, yeah, I could imagine cljdoc hosting an array of specifically defined examples. Like an example could include dependencies and snippets of some kind. The tricky bit here is defining examples in a data-driven way that is “portable” and gives sufficient information to run an example wherever. 

```
cljdoc run-ex some-I’d
```

Could for instance open a well enough defined example in a user friendly REPL like rebel-readline or similar. 

I think having those examples defined, or at least exportable as, data is critical to use them in a variety of environments.
